File: client/src/api/jobs.js
import api from './client.js';

/** Snapshot of all known jobs (active first, then recent history). */
export const listJobs = () => api.get('jobs').json();

/** Detail of a single job, including its full event log. */
export const getJob = (jobId) => api.get(`jobs/${jobId}`).json();

/** Cancel a queued or running job. No-op if already terminal. */
export const cancelJob = (jobId) => api.post(`jobs/${jobId}/cancel`).json();

/**
 * Subscribe to cross-job state updates as Server-Sent Events. Each message is
 * a `{type: 'jobs_state', jobs: [...]}` snapshot. Returns the EventSource so
 * the caller can `close()` it.
 *
 * This is the ONLY job stream the frontend opens. Per-job streams were
 * removed when we consolidated to a single SSE connection — opening one
 * EventSource per running job ate browser HTTP/1.1 connection-pool slots
 * and starved JS chunk + activity polling fetches under load.
 */
export function subscribeToJobsState(onSnapshot) {
  const es = new EventSource('/api/jobs/stream');
  es.onmessage = (msg) => {
    try {
      onSnapshot(JSON.parse(msg.data));
    } catch (e) {
      console.warn('[jobs] bad snapshot:', msg.data, e);
    }
  };
  es.onerror = (e) => {
    // EventSource auto-reconnects; we just log so a flaky connection is visible.
    console.debug('[jobs] state stream error (will retry)', e);
  };
  return es;
}

File: server-python/app/routers/jobs.py
"""Universal job endpoints: list, attach via SSE, cancel.

These work for any job kind (narrative, chub-import, meta, title, ...). The
frontend's toast bar + Activity view consume `/api/jobs` (live state via SSE
on `/api/jobs/stream`) and attach to a specific job's event stream via
`/api/jobs/{id}/stream`.
"""
from __future__ import annotations

import asyncio
import json
import logging

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from app.services import job_manager

router = APIRouter()
log = logging.getLogger("pendrift.jobs")


_SSE_HEADERS = {
    "Cache-Control": "no-cache",
    "Connection": "keep-alive",
    "X-Accel-Buffering": "no",
}


def _sse(event: dict) -> str:
    return f"data: {json.dumps(event, ensure_ascii=False)}\n\n"


@router.get("")
async def list_jobs():
    """All known jobs — active (queued + running) first, then recent history."""
    return {"jobs": job_manager.list_jobs()}


@router.get("/stream")
async def stream_state():
    """Long-lived SSE that pushes a `jobs_state` snapshot every time anything
    changes (creation, status transition, cancellation, completion). The first
    message is the current snapshot so the client doesn't need a separate GET.
    """
    queue = job_manager.subscribe_to_state()

    async def gen():
        try:
            yield ": jobs state stream open\n\n"
            while True:
                ev = await queue.get()
                yield _sse(ev)
        except asyncio.CancelledError:
            log.debug("jobs state subscriber disconnected")
            raise
        finally:
            job_manager.unsubscribe_from_state(queue)

    return StreamingResponse(gen(), media_type="text/event-stream", headers=_SSE_HEADERS)


@router.get("/{job_id}/stream")
async def stream_job(job_id: str):
    """Attach to a specific job's event stream. Replays everything emitted so
    far, then continues live. Closes when the job is terminal.

    Safe to disconnect and reconnect — the job runs in the background and the
    next attach will replay from the start."""
    job = job_manager.get_job(job_id)
    if not job:
        raise HTTPException(404, "Job not found")

    async def gen():
        yield ": job stream open\n\n"
        async for ev in job_manager.stream_job_events(job):
            yield _sse(ev)

    return StreamingResponse(gen(), media_type="text/event-stream", headers=_SSE_HEADERS)


@router.get("/{job_id}")
async def get_job(job_id: str):
    """Snapshot of a single job, including its full event log. Useful for the
    Activity detail panel without keeping an SSE open."""
    job = job_manager.get_job(job_id)
    if not job:
        raise HTTPException(404, "Job not found")
    return {**job.to_dict(), "events": job.events, "result": job.result}


@router.post("/{job_id}/cancel")
async def cancel_job(job_id: str):
    """Cancel a queued or running job. No-op if the job is already terminal."""
    if not job_manager.cancel_job(job_id):
        raise HTTPException(404, "Job not found or already finished")
    return {"ok": True}
